### ISSUE TITLE
Added Latency

### ISSUE THREAD

**Issue by EatonZ** [GoogleCloudPlatform/esp-v2#237]
I have ESPv2 deployed as a Cloud Run service, and the backend is another Cloud Run service hosting my simple API.
Both are in us-central1.
Doing a quick TTFB test on my API through the ESPv2 Cloud Run URL yields an average of **68ms**.
Doing the same test on the API directly (essentially bypassing ESPv2) yields an average of **56ms**.

Is this amount of latency normal?
I recognize this project is still in beta - is there ongoing work being done to reduce latency before the GA release?

If you need more details from me, let me know. I'd be interested in seeing your results as well.

**Comment by nareddyt** [comment]
Within ESPv2, the latency will drastically change based on what [filters](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/doc/architecture.md) are enabled. We automatically enable filters based on the API configuration you provide (OpenAPI spec or gRPC service config). Have you added any features such as API key checks, JWT token verification, gRPC transcoding, etc? Or do you expect your request to essentially pass through without any extra modifications / checks?

ESPv2 also heavily uses caching (eviction time ~5 minutes). Make sure run the same request multiple times and then measure the latency.

There are some items outside of ESPv2 control. When deploying ESPv2 on Cloud Run, you are adding in an extra network hop before the request hits your backend server. This is inside Google Cloud's data centers, so it shouldn't be too bad, but it may add a few ms. This is an inherent problem in a serverless deployment.

If you have a strong requirement for low latency, then you may want to try using [Cloud Endpoints for GKE](https://cloud.google.com/endpoints/docs/openapi/get-started-kubernetes-engine) and compare the results.

Unfortunately we do not have performance benchmarks that test with complex configuration. For some of our simple E2E tests on serverless, I only see a latency overhead of 6ms for 95th percentile of traffic.

**Comment by qiwzhang** [comment]
You can also try Cloud Run for [Anthos](https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run-anthos).  ESPv2 is running at the same cluster as your backend.  It network latency between ESPv2 and your backend should be much smaller.

**Comment by EatonZ** [comment]
Thank you for the responses!

@nareddyt:

> Have you added any features such as API key checks, JWT token verification, gRPC transcoding, etc? Or do you expect your request to essentially pass through without any extra modifications / checks?

None of that - it's just a simple "hello world" JSON API. Do I need to try and explicitly disable anything?

> Make sure run the same request multiple times and then measure the latency.

I made sure to do this to get consistent results, and so the cold startup would not skew the results.

> When deploying ESPv2 on Cloud Run, you are adding in an extra network hop before the request hits your backend server.

Indeed - I know it's expected to have a little bit of extra latency due to the nature of this service. Just wanted to know how much is typically expected.

> For some of our simple E2E tests on serverless, I only see a latency overhead of 6ms for 95th percentile of traffic.

I'm seeing more than that for my simple app. Could different regions have better performance?

**Comment by EatonZ** [comment]
@qiwzhang Thanks for the suggestion. I don't think I will be looking at Anthos right now, but I'll keep it in mind.

**Comment by EatonZ** [comment]
Also, to remove my Internet speed as a factor, here's numbers I see in both services' Cloud Run logs:
The backend service completes this particular request in **3-5ms**.
The ESPv2 service completes it in **15-25ms**.
I am looking at the `latency` field of [httpRequest](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#httprequest) as seen in the Logs Viewer for Cloud Run logs.

A logical next step: do some performance profiling of ESPv2. My thought is to have some sort of logging that will say how long the request to the backend takes. If that takes up the majority of the final ms, then that would help confirm whether networking is the bottleneck instead of ESPv2 code overhead. Is this something that can be done?

**Comment by qiwzhang** [comment]
For latency debugging,  it is easier to use trace. You can look at the
trace data in the cloud console.  By default,  ESPv2 is using 1/1000
sampling rate to enable trace.   But you can adjust it in the flag
<https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/generic/start_proxy.py#L443>
.

**Comment by EatonZ** [comment]
Hi @qiwzhang, here's the result (host redacted for privacy):

![Trace](https://user-images.githubusercontent.com/3020941/88345954-2828ef00-cd15-11ea-9b96-cb288d644ee4.png)

I haven't use Trace before, so correct me if I am wrong:
-28.571 ms is the total time it takes for the backend request + ESPv2 overhead
-28.285 ms is the backend request time
-0.285 ms is the ESPv2 overhead

If I am right, it looks like ESPv2 overhead is quite minimal. Just doing a few tests, I have seen the latency vary wildly. I have seen it go as high as 85 ms. That kind of latency is pretty high - what can be done to improve it?

**Comment by qiwzhang** [comment]
Yes, your understanding is correct.   So the next step is to capture the
trace for the high latency requests, like the one with 85ms, to see where
85ms spend.  You can turn on trace for all requests by changing the
sampling rate.

**Comment by EatonZ** [comment]
> You can turn on trace for all requests by changing the sampling rate.

I already set `--tracing_sample_rate` for ESPv2, where else should I be changing it?

**Comment by qiwzhang** [comment]
if you set it to 1.0,  all requests should have a trace.   Is that not the
case?

**Comment by EatonZ** [comment]
I see traces for every request, so yes it is working.

I meant to say that the 85ms request was indeed traced, as that is how I observed that number.
That particular trace indicates a similar ESPv2 overhead, just a much higher network response time:

![Trace2](https://user-images.githubusercontent.com/3020941/88441443-24fa3580-cddf-11ea-87a6-129b19ac8fba.png)

**Comment by qiwzhang** [comment]
"router" is the latency from Envoy upstream,  it means the backend.   Now
we want to know how much of it is from the network latency and how much is
from the backend.
 Since your backend doesn't support trace, we don't know how much time is
spent on your backend.
Maybe you can chain two ESPv2 services,  one after another,  then the trace
will show two "ingress" and "router",  it will show the latency from one
Cloud Run service to another

**Comment by nareddyt** [comment]
+1 to wayne's comment. ESPv2 creates the "router" trace at the time that the request leaves ESPv2 and heads to your backend. It also contains the network delay that it takes for the request to reach your backend, as well as the delay for the response to go from the backend to ESPv2 again.

So it definitely seems like the majority of the latency ESPv2 introduces is coming from the network. Some options to narrow that down:
- Wayne's comment
- Add stackdriver trace to your backend
- Try to correlate timestamps in your backend logs to the timestamps of this trace. It will be difficult, we're talking about a few ms of difference.

**Comment by EatonZ** [comment]
I'll take a closer look and report back.

Something else I wanted to ask, does ESPv2 keep the connection to the backend open (Keep-Alive) for use in subsequent requests, or is a new connection opened for each request that comes in?

**Comment by qiwzhang** [comment]
Envoy upstream is http1.1, so the connections to the backend are re-used.

**Comment by EatonZ** [comment]
@qiwzhang I put another ESPv2 instance in front. So it goes Request -> Gateway 2 -> Gateway 1 -> Backend
It looks like the network latency is around 20ms. It's 20ms from Gateway 2 to Gateway 1, then another 20ms to the Backend.
Here's screenshots:

![Trace 3](https://user-images.githubusercontent.com/3020941/88450281-dae37500-ce1b-11ea-8a1d-80e24ab0e800.png)
![Trace 4](https://user-images.githubusercontent.com/3020941/88450282-dfa82900-ce1b-11ea-8bce-04ddc3023dcb.png)

**Comment by EatonZ** [comment]
By the way, I didn't need to set `--tracing_sample_rate` at all. Turns out all requests are being traced by default. Not sure if that's intended, but that is what I am seeing.🤔

**Comment by nareddyt** [comment]
Thanks for the traces, I see the 20ms of latency added for your request.

I created PR #247 to add more tracing for our backends in our e2e tests. Once this runs for a day, I'll have more specific numbers for the network latencies. Then I can judge if 20ms is the p50 we see as well, i'll update you in a day or so.

(Unrelated: But I'm surprised you had 2 different traces for a single request. They should be merged into 1 parent with multiple nested children, take a look at the image in PR #247. Perhaps you need to set ESPv2 flag `--tracing_outgoing_context=x-cloud-trace-context` like I did? It really depends on what library your backend uses.)

Your last comment about `--tracing_sample_rate` is surprising, the default is only 0.1% of requests being traced. I can double check later, but I've definitely noticed fewer traces when not setting it. How did you set the flag exactly in ESPv2, it's possible you forgot to remove it during a redeploy?

**Comment by EatonZ** [comment]
> Perhaps you need to set ESPv2 flag --tracing_outgoing_context=x-cloud-trace-context like I did?

I did not have that set. That could be why.

> Your last comment about --tracing_sample_rate is surprising, the default is only 0.1% of requests being traced. I can double check later, but I've definitely noticed fewer traces when not setting it. How did you set the flag exactly in ESPv2, it's possible you forgot to remove it during a redeploy?

I just checked this again. No Cloud Run environment variables are present, and the Cloud Run envoy log says `TracingSamplingRate:0.001`.
Sending 10 requests to the gateway, all of them show up in Trace List. I did not add a `x-cloud-trace-context` header either.

> i'll update you in a day or so.

Looking forward to seeing your results.🙂

**Comment by nareddyt** [comment]
Quick update: The code I added in PR #247 does not give a comprehensive view of latency, it does not correlate traces correctly.

I'm waiting on https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/116 and the release of this new package. Once that's done, I should be able to get better understanding of the networking latency for various Google Cloud Backends (cloud function, cloud run, GKE, etc).

Will probably take a few weeks.

**Comment by EatonZ** [comment]
Ok, I will await your update.

**Comment by EatonZ** [comment]
@nareddyt Version 0.4.0 of that package was released today and includes that feature you need.

**Comment by nareddyt** [comment]
Awesome, thanks for the reminder. I see it's [published here](https://www.npmjs.com/package/@google-cloud/opentelemetry-cloud-trace-propagator). I'll try to update our backend this week.

**Comment by nareddyt** [comment]
Almost works, but there is a small bug that prevents context propagation. Waiting on this PR https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/147 to be published and released again.

**Comment by nareddyt** [comment]
Thanks! I was playing around with it on my latest commits in #252. Have some other higher-priority bugs to solve first, then I'll test it out.

**Comment by nareddyt** [comment]
Ran into more issues: https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/161

I will keep checking back and make incremental progress on this when I have time :)

**Comment by nareddyt** [comment]
I did some preliminary analysis on the latency. I'll can spend time later getting actual p50/90/95 metrics and writing up a doc in this repo, but I'll describe the most common latencies I'm seeing.

### Deployment

- ESPv2 on Cloud Run and my backend on Cloud Run. Both are in `us-central1`.
- Backend is a simple HTTP nodejs backend using `express-js` for the web server. It has some modules for body parsing and tracing.
- ESPv2 is configured to do the bare minimum processing. It receives the HTTP request and proxies it to the backend. No JWT authentication, no API key checks, no gRPC transcoding, etc.

### Requests

- 1 GET request every 5 seconds with no body. 
- Backend echos back all the headers it receives in a JSON response body.

### Overall latency

This counts the total latency from ESPv2 receiving the request to sending back a response. It doesn't account for the latency the client sees.
- Majority of requests are under 40ms
- Median latency seems to be 25ms

![image](https://user-images.githubusercontent.com/11142171/92664059-fd98f280-f2d0-11ea-96c1-f0dafad9677d.png)

### Latency breakdown

- The request spends very little time inside ESPv2 (as expected, ESPv2 is not doing any checks in this deployment). In most cases, < 0.5 ms.
- The majority of the latency is the request being routed to the Cloud Run backend. On average, < 17ms of networking delay here.
- Backend takes 1ms.
- There is more latency for the response to reach ESPv2 from the backend. On average, < 6ms of networking delay here.
- The response spends very little time inside ESPv2 (as expected). On average, < 0.2 ms.

Here is one of the most common traces I see.

![image](https://user-images.githubusercontent.com/11142171/92664798-f672e400-f2d2-11ea-84de-f7a8d62ef38d.png)

### Verification

- The overall latency observed here matches the latency that is recorded in the endpoints log.
- It also matches the endpoints service graphs. We can see the p50 matches the trace example I gave above.

![image](https://user-images.githubusercontent.com/11142171/92665267-1f47a900-f2d4-11ea-9232-15c8b92303b1.png)

### Final comments

@EatonZ this matches the traces you posted in the [comment here](https://github.com/GoogleCloudPlatform/esp-v2/issues/237#issuecomment-663816267). Given that the networking hop seems to take 24ms, then your numbers of the total client seeing "an average of 56ms" is certainly possible. I will admit that this is a little is higher than I expected, but there's not much we can do about it.

If this is too high of a latency for you, then you can try using [ESPv2 on GKE](https://cloud.google.com/endpoints/docs/openapi/get-started-kubernetes-engine-espv2). We Beta launched this a couple of weeks ago. Of course, serverless is a lot simpler to manage than kubernetes :)

Since this matches your observations, I'll close this issue now. We'll try to prioritize setting up performance benchmarks and a automated latency reports next quarter, some of our recent tracing changes make it doable.

**Comment by EatonZ** [comment]
@nareddyt Thank you for your final report. It looks like you basically confirmed what I suspected, that the networking hop is the bulk of the added latency.

> I will admit that this is a little is higher than I expected, but there's not much we can do about it.

Indeed, this seems like a topic for your datacenter networking team and not ESPv2.
I am on a fiber connection at home and sometimes I can reach Cloud Run services (no ESPv2 gateway) in 35ms.
Since ESPv2 -> backend Cloud Run service is all done in the same datacenter, I expected it to be pretty quick. I don't have anything to compare it to right now, but I personally would love if the latency was just single-digit.

> If this is too high of a latency for you, then you can try using ESPv2 on GKE.

Maybe in the future, but not at this time.

Since there doesn't seem to be a fault in ESPv2, would it be possible for you to open a ticket with the networking team or something? I imagine there is no quick fix here, but it seems like it's worth more discussion.

**Comment by nareddyt** [comment]
We had some internal discussion between Cloud Endpoints and Cloud API Gateway. We did agree we need to better understand the sources of latency, this is likely something we'll start working on next quarter.

**Comment by EatonZ** [comment]
Hi @nareddyt, it's been some time, has there been any improvements regarding latency? Maybe consider running your analysis again to see if anything is better now.

**Comment by nareddyt** [comment]
Hi @EatonZ. I don't have time this week to do a full analysis again, but I will try next week.

Looking at trace data from our Serverless end-to-end tests shows the p50 total request time is lower than 25 ms. I am hopefully it is reduced, but there may be other factors that are not accounted for.

![image](https://user-images.githubusercontent.com/11142171/111186753-b6463400-8589-11eb-9d35-bb5ad74dcf99.png)

**Comment by EatonZ** [comment]
Hi @nareddyt, let me know when you do the full analysis.🙂
